src/components/common/RestEntity/README.md

```diff
@@ -0,0 +1,882 @@
+
+That's where the sub-components come in.
+
+We have the `Singleton` for success state, `Error` for error state, and `Loading` for loading state.
+
+
+Let's think all the way back to the original `ProgramComponent.js`. What did its return statement look like?
+
+```javascript
+return (
+    <>
+    {
+      // if data exists and is truthy (i.e., not `undefined` or `null`)
+      // then we'll render our program card with information!
+      data ?
+        <div>
+          <h1>Program Name: {data.programName}</h1>
+          <p>Description: {data.description}</p>
+        </div>
+        // otherwise, we'll show the user that we're loading up their data
+        : <div>Loading...</div>
+    }
+    {
+      // if error exists, then we'll render this warning to our user!
+      error && <p>Oh no! There's been a problem!</p>
+    }
+    </p>
+  );
+```
+
+It was handling a bunch of logic to decide whether it was in a place to show data or not.
+
+But the essence behind all the ternaries and truthiness fail-safes contained three different
+views based on the three shapes of state (LOADING, SUCCESS, ERROR).
+
+LOADING:
+
+```javascript
+// if data is still being fetched
+// we'll show the user that we're loading up their data
+return <div>Loading...</div>;
+```
+
+ERROR:
+
+```javascript
+// if we hit an error, then we'll render this warning to our user!
+return <p>Oh no! There's been a problem!</p>;
+```
+
+SUCCESS:
+
+```javascript
+// if data was successfully fetched, 
+// then we'll render our program JSX to display that data
+return  (
+  <div>
+    <h1>Program Name: {data.programName}</h1>
+    <p>Description: {data.description}</p>
+  </div>
+);
+```
+
+Do you notice anything interesting there? Of the three different states, only ONE depends on external data of any sort—SUCCESS. We can display anything we want to for a Loading message or an Error message... the only thing that depends on the actual data is the success!! And logically so, because if you're loading data or failed to get the data, then you definitely don't have any data that you could depend on.
+
+So let's plug our `Error` and `Loading` states into our new and improved `ProgramComponent.js`
+
+```javascript
+import React from "react";
+import RestEntity from "./RestEntity";
+
+const ProgramComponent = props => {
+  return (
+    <RestEntity href={`/programs/program/${props.programId}`}>
+
+      {/* WE STILL DON'T KNOW WHAT TO DO HERE */}
+      <RestEntity.Singleton />
+      
+      {/* The JSX inside this will only display if we hit an error */}
+      <RestEntity.Error>
+        <p>Oh no! There's been a problem!</p>
+      </RestEntity.Error>
+      
+      {/* The JSX inside this will only display when we're loading up data */}
+      <RestEntity.Loading>
+        <div>Loading...</div>
+      </RestEntity.Loading>
+
+    </RestEntity>
+  );
+};
+
+export default ProgramComponent;
+```
+
+Wow! We now have a component that can:
```

Comment:
This structure is good, also should probably be moved somewhere to the top too for a clear goal in mind, although I realize why you approached it the way that you did.